sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java

```diff
@@ -136,77 +195,86 @@ public Write withDatabaseId(String databaseId) {
-   *
-   * <p>Commits are non-transactional.  If a commit fails, it will be retried (up to
-   * {@link SpannerWriterFn#MAX_RETRIES} times). This means that the mutation operation should be
-   * idempotent.
-   *
-   * <p>See <a href="https://cloud.google.com/spanner">Google Cloud Spanner documentation</a>.
-   */
+  /** Batches together and writes mutations to Google Cloud Spanner. */
   @VisibleForTesting
-  static class SpannerWriterFn extends DoFn<Mutation, Void> {
-    private static final Logger LOG = LoggerFactory.getLogger(SpannerWriterFn.class);
+  static class SpannerWriteFn extends DoFn<Mutation, Void> {
+    private static final Logger LOG = LoggerFactory.getLogger(SpannerWriteFn.class);
+    private final Write spec;
     private transient Spanner spanner;
-    private final String instanceId;
-    private final String databaseId;
-    private final int batchSize;
     private transient DatabaseClient dbClient;
     // Current batch of mutations to be written.
-    private final List<Mutation> mutations = new ArrayList<>();
+    private List<Mutation> mutations;
+    private long batchSizeBytes = 0;
 
     private static final int MAX_RETRIES = 5;
     private static final FluentBackoff BUNDLE_WRITE_BACKOFF =
         FluentBackoff.DEFAULT
-            .withMaxRetries(MAX_RETRIES).withInitialBackoff(Duration.standardSeconds(5));
+            .withMaxRetries(MAX_RETRIES)
+            .withInitialBackoff(Duration.standardSeconds(5));
 
     @VisibleForTesting
-    SpannerWriterFn(String instanceId, String databaseId, int batchSize) {
-      this.instanceId = checkNotNull(instanceId, "instanceId");
-      this.databaseId = checkNotNull(databaseId, "databaseId");
-      this.batchSize = batchSize;
+    SpannerWriteFn(Write spec) {
+      this.spec = spec;
     }
 
     @Setup
     public void setup() throws Exception {
-        SpannerOptions options = SpannerOptions.newBuilder().build();
-        spanner = options.getService();
-        dbClient = spanner.getDatabaseClient(
-            DatabaseId.of(options.getProjectId(), instanceId, databaseId));
+      spanner = spec.getSpannerOptions().getService();
+      dbClient =
+          spanner.getDatabaseClient(
+              DatabaseId.of(projectId(), spec.getInstanceId(), spec.getDatabaseId()));
+      mutations = new ArrayList<>();
+      batchSizeBytes = 0;
     }
 
     @ProcessElement
     public void processElement(ProcessContext c) throws Exception {
       Mutation m = c.element();
       mutations.add(m);
-      int columnCount = m.asMap().size();
-      if ((mutations.size() + 1) * columnCount >= batchSize) {
+      batchSizeBytes += MutationSizeEstimator.sizeOf(m);
+      if (batchSizeBytes >= spec.getBatchSizeBytes()) {
         flushBatch();
       }
     }
 
+    private String projectId() {
+      return spec.getProjectId() == null
+          ? ServiceOptions.getDefaultProjectId()
```

Comment:
I wanted to swap to use ServiceOptions.getDefaultProjectId() a while ago and have not been able to perform that migration for GcpOptions. If we can include it here great. This would remove all our hand-coded detect default project code. Note that by doing this we would remove support for gcloud users who have old gcloud SDK installations.

I requested the original change in gcloud-java-core to have this exposed:
https://github.com/GoogleCloudPlatform/google-cloud-java/pull/1380